storage/storage_dest.go

```diff
@@ -77,18 +78,20 @@ type storageImageDestination struct {
 	indexToStorageID map[int]*string
 	// All accesses to below data are protected by `lock` which is made
 	// *explicit* in the code.
-	uncompressedOrTocDigest map[digest.Digest]digest.Digest                       // Mapping from layer blobsums to their corresponding DiffIDs or TOC IDs.
-	fileSizes               map[digest.Digest]int64                               // Mapping from layer blobsums to their sizes
-	filenames               map[digest.Digest]string                              // Mapping from layer blobsums to names of files we used to hold them
-	currentIndex            int                                                   // The index of the layer to be committed (i.e., lower indices have already been committed)
-	indexToAddedLayerInfo   map[int]addedLayerInfo                                // Mapping from layer (by index) to blob to add to the image
-	blobAdditionalLayer     map[digest.Digest]storage.AdditionalLayer             // Mapping from layer blobsums to their corresponding additional layer
-	diffOutputs             map[digest.Digest]*graphdriver.DriverWithDifferOutput // Mapping from digest to differ output
+	blobDiffIDs           map[digest.Digest]digest.Digest                       // Mapping from layer blobsums to their corresponding DiffIDs
+	tocDigests            map[digest.Digest]digest.Digest                       // Mapping from layer blobsums to their corresponding TOC Digest
```

Comment:
If we don’t validate the blob sums (compressed digests) ourselves, we can see conflicts in malicious inputs.

Reply:
the TOCs are always validated by `store.ApplyDiffWithDiffer()`.

If a layer has an annotation referring to a TOC, then we use that for the lookup in the storage (where the TOC are stored only after being validated).  This is conceptually the same thing we do with the diffID, if an image asks for a certain diffID, we lookup in the storage if we know that layer

Reply:
Yes, but the map _key_ is not validated on the TOC path.

On the non-TOC path, same compressed digest implies same uncompressed digest, and we do compute both of them, so we trust the values.

On the TOC path, we do verify the TOC digest (map value), but the key is currently the compressed digest. I don’t see that anything prevents the attacker to create an image with
```
{digest = $same, tocDigest = $toc1}
{digest = $same, tocDigest = $toc2}
{digest = $same, tocDigest unset}
```
and what happens then? My instinct is to make this impossible instead of trying to reason through it.

(One possible way to handle this concern would be to reject images with conflicting data for the same compressedDigest values. But, see elsewhere, I think we need to maintain “how was this pulled” data by layer index to compute correct image IDs anyway… so using layer indices for map keys here seems ultimately simpler to me, admittedly without having tried to implement it.)

Reply:
I've not thought of all possible implications but while compressed and uncompressed digests can always be used  interchangeably (one compressed digest refers to one uncompressed digest, and vice-versa), that is not true for TOC digests.  e.g. it is enough to add some padding to the tarball to cause compressed and uncompressed digest to change, but that won't affect the TOC digest.

Reply:
- Actually, a single uncompressed digest can have many compressed-digest values (one per algorithm is the lower bound, and most compression algorithms are not deterministic)
- It’s not “quite true” for uncompressed digests, in that sense, either — a tar file can have “the same contents” but a different ending padding (to mention a recent example).

Here, I’m not worried about finding all possible duplicates and deduplicating. That’s an optimization.

I am worried about _incorrectly_ treating two things as duplicates when we don’t know for sure they are identical. That’s a vulnerability.

Reply:
Resolved in #2288.